Fix theme for cursors and the inside check on checkboxes.

jpc-ae opened the pull request:
## Description

This pull relates to temporary fixes removing some hard-coded default values which make theming more difficult. The changes comprise: text cursor (caret) and selection handle are both set to the primary color per the images and implementations in the material design specs, and the drawn check in checkboxes now default to the background color, as generally an icon with a transparent hole should be used there instead.

## Related Issues

#17635, #56082, #32023

jpc-ae commented:
Not sure why the golden tests are failing for text_field_test.dart. Perhaps someone with more experience on this project can spot where the issue is?

jpc-ae commented:
After updating goldens, flutter test passes all checks.

HansMuller commented:
It looks like you may have applied dartfmt to the sources, see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#formatting. Please limit this PR's changes to the minimum needed to achieve the goals set out in the PR's description. That said, we're in the middle of evolving Flutter's theme system, see [flutter.dev/go/material-theme-system-updates](https://flutter.dev/go/material-theme-system-updates). Eventually, CheckBox should be configurable with its own theme and its defaults should be based on the overall Theme's ColorScheme.

jpc-ae commented:
You are correct. I forgot to disable it from autorunning on save. I have manually reverted the changes made by dartfmt, so it should be conformant now.

I also understand checkboxes, radio buttons and switches all need to be updated to the current guidelines with theming in mind, and have started taking a stab at that. But I felt this temporary fix was warranted.

jpc-ae commented:
Alright, just checked the diff and it seems indentation is still being messed with, so I will go over and try to manually revert that as well.

jpc-ae commented:
There we are, only relevant changes made.

Piinks commented on packages/flutter/lib/src/material/checkbox.dart:
```diff
@@ -135,7 +135,7 @@ class Checkbox extends StatefulWidget {
   /// The color to use for the check icon when this checkbox is checked.
   ///
-  /// Defaults to Color(0xFFFFFFFF)
+  /// Defaults to {ThemeData.canvasColor]
```
Nit:
```suggestion
  /// Defaults to [ThemeData.canvasColor]
```

Piinks reviewed (commented):
Hi @jpc-ae welcome!

jpc-ae commented:
> Hi @jpc-ae welcome!

Hi, thanks for the nit!

Piinks commented on packages/flutter/lib/src/material/theme_data.dart:
```diff
@@ -312,8 +312,8 @@ class ThemeData with Diagnosticable {
     // TODO(hansmuller): We need a TextFieldTheme to handle this instead, https://github.com/flutter/flutter/issues/56082
-    cursorColor = cursorColor ?? const Color.fromRGBO(66, 133, 244, 1.0);
-    textSelectionHandleColor ??= isDark ? Colors.tealAccent[400] : primarySwatch[300];
+    cursorColor ??= isDark ? Colors.tealAccent[200] : primarySwatch;
+    textSelectionHandleColor ??= isDark ? Colors.tealAccent[200] : primarySwatch;
```
Can you provide a reference for the color choices you have made throughout this change? Did they come from the material spec?

jpc-ae replied:
The color choices generally come from platform-specific material specs.

For the checkColor, in making it themable, I chose canvasColor as the default check color is the same as "?attr/colorSurface" according to what I understood from this page: https://developer.android.com/reference/com/google/android/material/checkbox/MaterialCheckBox and by looking at the many examples in the design spec (https://material.io/components/selection-controls). In fact, there doesn't seem to be a provision for theming this separately on android and the control might benefit from animating the check as a cutout rather than an overlay. However, this would be a more comprehensive change in connection with the Material Theme overhall being worked on, so I made this small change as a stopgap.

As for the text cursor, this should default to primary according to both the android and iOS material specs for text inputs (Links from here: https://material.io/components/text-fields). While there is no specific documentation on the selection handle color, this always matches the collapsed cursor color in tests. This is especially obvious on iOS where the handle extends upwards to bookend selections. Both according to the spec and as confirmed on #56082, this is designed to match the colorControlActivated color used for focused text field borders. Again, I created this pull mainly as a stopgap.

As for the dark theme colors, I updated everything to match the accentColor (teal[200]) for consistency since this is the current default for the textSelectionHandleColor. These should arguably be the same as primary, however the current dark theme primary is not a color despite Material dark theme guidelines (https://material.io/design/color/dark-theme.html#anatomy) suggesting it should be (a desaturated version of light primary). Changing the entire default dark theme to better match the guidelines was out of scope for this pull which is just about clearing out unthemable colors.

Piinks reviewed (commented):
It looks like most of the pre-submit checks are failing here due to a versioning issue. This happens sometimes when branches age away from the master branch. Can you update your branch with the latest changes from master? That should resolve the failures.

I think @darrenaustin is already working on TextField/cursor theming, will defer to him on that here.
We may just reduce the scope of this to include the checkbox updates.

jpc-ae committed:
Remove hardcoded color from checkbox check and text field cursor

jpc-ae commented:
I have now updated this branch to be even with master. There were several conflicts due to the new world of buttons, so I reset and introduced the changes on top in a clean way. Hope that is okay.

Piinks commented:
@darrenaustin can you take a look at this change re: theming?

darrenaustin reviewed (commented):
Thanks so much for the contribution! Sorry I didn't get to this sooner. 

As @Piinks already mentioned the change for the cursor color is already being handled by a migration to a new [`TextSelectionTheme` ](https://github.com/flutter/flutter/pull/62014), so that part of this PR will not be needed.

As for the change in the checkbox `checkColor`, this is a breaking change as there are clients that depend on the default being white. There are internal google tests that will fail with this change. If we want to proceed with changing this we will need to have an opt-in transition as described in the [breaking change section](https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes) of the hygiene doc. 

Given that that is a somewhat involved process, we might want to take this opportunity to convert the checkbox over to using its own `CheckboxTheme` as outlined in the [theme system update doc](flutter.dev/go/material-theme-system-updates), instead of just changing this one default.

If this is something you would like to work on, we would be more than willing to assist you. If not hopefully someone on the team will be able to get to this soon.
